runtime-modules/bounty/src/lib.rs

```diff
@@ -0,0 +1,676 @@
+        NotBountyCreator,
+
+        /// Work period cannot be zero.
+        WorkPeriodCannotBeZero,
+
+        /// Judging period cannot be zero.
+        JudgingPeriodCannotBeZero,
+
+        /// Invalid bounty stage for the operation.
+        InvalidBountyStage,
+
+        /// Insufficient balance for a bounty cherry.
+        InsufficientBalanceForBounty,
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        /// Predefined errors
+        type Error = Error<T>;
+
+        /// Emits an event. Default substrate implementation.
+        fn deposit_event() = default;
+
+        /// Creates a bounty. Metadata stored in the transaction log but discarded after that.
+        /// <weight>
+        ///
+        /// ## Weight
+        /// `O (W)` where:
+        /// - `W` is the _metadata length.
+        /// - DB:
+        ///    - O(1)
+        /// # </weight>
+        #[weight = WeightInfoBounty::<T>::create_bounty_by_member()
+              .max(WeightInfoBounty::<T>::create_bounty_by_council())]
+        pub fn create_bounty(origin, params: BountyCreationParameters<T>, _metadata: Vec<u8>) {
+            let bounty_creator_manager = BountyCreatorManager::<T>::get_bounty_creator(
+                origin,
+                params.creator.clone()
+            )?;
+
+            bounty_creator_manager.validate_balance_sufficiency(params.cherry, params.creator_funding)?;
+
+            Self::ensure_create_bounty_parameters_valid(&params)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            bounty_creator_manager.slash_balance(params.cherry, params.creator_funding);
+
+            let next_bounty_count_value = Self::bounty_count() + 1;
+            let bounty_id = T::BountyId::from(next_bounty_count_value);
+
+            let bounty = Bounty::<T> {
+                total_funding: params.creator_funding,
+                creation_params: params.clone(),
+                state: BountyMilestone::Created(Self::current_block()),
+            };
+
+            <Bounties<T>>::insert(bounty_id, bounty);
+            BountyCount::mutate(|count| {
+                *count = next_bounty_count_value
+            });
+            Self::deposit_event(RawEvent::BountyCreated(bounty_id, params));
+        }
+
+        /// Cancels a bounty.
+        /// # <weight>
+        ///
+        /// ## weight
+        /// `O (1)`
+        /// - db:
+        ///    - `O(1)` doesn't depend on the state or parameters
+        /// # </weight>
+        #[weight = WeightInfoBounty::<T>::cancel_bounty_by_member()
+              .max(WeightInfoBounty::<T>::cancel_bounty_by_council())]
+        pub fn cancel_bounty(origin, creator: BountyCreator<MemberId<T>>, bounty_id: T::BountyId) {
+            let bounty_creator_manager = BountyCreatorManager::<T>::get_bounty_creator(
+                origin,
+                creator.clone(),
+            )?;
+
+            ensure!(
```

Comment:
1. we will need to safely recover bounties in multiple places, so lets refactor this check+get into `ensure_bounty_exists`.
2. Let us avoid this mutable, this seems avoidable at little cost.